Async test result

Phuong-Le-NN opened the pull request:
# Description

<!-- Describe the context / changes in this PR -->
Async test result when one buddy click Run button of Leetcode

## Screenshots

<!-- For UI changes, include screenshots / recordings and describe the expected user flow -->
<img width="1456" height="1380" alt="image" src="https://github.com/user-attachments/assets/0fd44d14-58e5-4ea0-9d2d-da384a9f898a" />

## Test
Click Run code on one buddy and wait a bit to see the Test Result tab filled out on the other.

## Checklist

If you're making changes to the extension, please run through the following checklist to make sure that we don't have
any regressions. Note that we plan to add integration tests in the future!

- [ ] Create room and join room on at least 2 browsers
- [ ] Ensure that code and tests are correctly stream
- [ ] Verify that when reloading, user can automatically join the room

## Possible Downsides

<!-- List anything we should be aware of -->

## Additional Documentations

<!-- Describe any documentations or references that would be helpful for us to review your code -->

nickbar01234 commented on extension/src/types/index.ts:
```diff
@@ -15,14 +15,28 @@ interface Assignment {
 }
 
+interface ResultAssignment {
+  input: Record<string, string>;
```
Can we perhaps reuse the same interface? https://github.com/nickbar01234/codebuddy/pull/549/files#diff-96d91a7cf927d8a78ffc02453d6987162a4f8d07c9ffe8c1269a3ee75c168652R13-R16. I also think some problems do not have variables, i.e OOP-related questions

nickbar01234 commented on extension/src/components/panel/editor/tab/TestResultTab.tsx:
```diff
@@ -0,0 +1,118 @@
+                    {selected ? (
+                      passed ? (
+                        <button className="bg-fill-3 dark:bg-dark-fill-3 hover:bg-fill-2 dark:hover:bg-dark-fill-2 hover:text-label-1 dark:hover:text-dark-label-1 text-label-1 dark:text-dark-label-1 relative inline-flex items-center whitespace-nowrap rounded-lg px-4 py-1 text-sm font-semibold focus:outline-none">
+                          V Case {idx + 1}
```
What is `V`? May be we can use an icon component instead?

Commit:
able to synce test result by intercepting fetch request after buddy run test cases

Commit:
commit to keep progress
changed type to reuse test case assignment interface
fix for case when input has no variable name

Commit:
not breaking non error test result cases but not working for error test result cases yet

Commit:
commit to keep progress
backend working?
frontend not yet for error cases

Commit:
commit to keep progress before starting on frontend, test result from backend compplete

Commit:
frontend in progress

Commit:
frontend almost complete + fixed backend a little t send the correct object for each error case

Commit:
completed and cleaned up on frontend
fix + refactor slightly on backend for grouping test result

Commit:
clean up and add comments

Commit:
add icon and fixed minor ui

nickbar01234 commented:
Can you comment with the latest screenshot of the new UI?

nickbar01234 commented on extension/src/utils/messages.ts:
```diff
@@ -17,6 +17,44 @@ export const getTestsPayload = (
 
+export const getTestResultsPayload = (
+  variables: Question["variables"] | undefined,
+  testResults?: any
```
May be without `?` too

Phuong-Le-NN commented:
<img width="705" height="743" alt="Screenshot 2026-04-01 at 1 15 54 PM" src="https://github.com/user-attachments/assets/0960ecc5-c4da-457f-a023-41beac3124b9" />
<img width="589" height="598" alt="Screenshot 2026-04-01 at 1 26 05 PM" src="https://github.com/user-attachments/assets/4d8ccaff-9b6e-4dab-8f57-a1e4658ec2b6" />
<img width="599" height="597" alt="Screenshot 2026-04-01 at 1 26 25 PM" src="https://github.com/user-attachments/assets/5030d3a9-aa07-4f7e-a1cc-d367e1cdfd6b" />
<img width="590" height="593" alt="Screenshot 2026-04-01 at 1 26 42 PM" src="https://github.com/user-attachments/assets/fc6831f8-0af9-44cf-92cb-c8564bb83bd6" />
<img width="587" height="602" alt="Screenshot 2026-04-01 at 1 27 55 PM" src="https://github.com/user-attachments/assets/08dfac01-1c75-483b-9fe0-28e147acf7f7" />
<img width="645" height="601" alt="Screenshot 2026-04-01 at 1 28 44 PM" src="https://github.com/user-attachments/assets/cc56fd31-982b-419a-97a6-3dc682d1f893" />
<img width="615" height="604" alt="Screenshot 2026-04-01 at 1 32 03 PM" src="https://github.com/user-attachments/assets/7b45aa51-637a-4cba-8a72-a79791278c8c" />

Phuong-Le-NN commented on extension/src/entrypoints/content.tsx:
```diff
@@ -11,6 +11,12 @@ export default defineContentScript({
+    s.src = chrome.runtime.getURL("proxy.js");
+    s.type = "text/javascript";
+    s.onload = () => s.remove();
+    document.documentElement.appendChild(s);
```
This creates a script element to load `proxy.js` into the page, then removes the tag. I realized this is redundant. I’ll remove it.

Phuong-Le-NN commented on extension/src/utils/messages.ts:
```diff
@@ -17,6 +17,44 @@ export const getTestsPayload = (
 
+export const getTestResultsPayload = (
+  variables: Question["variables"] | undefined,
+  testResults?: any
```
yess, here


Compile Error
```
{
    "status_code": 20,
    "lang": "cpp",
    "run_success": false,
    "compile_error": "Line 5: Char 5: error: non-void function does not return a value [-Werror,-Wreturn-type]",
    "full_compile_error": "Line 5: Char 5: error: non-void function does not return a value [-Werror,-Wreturn-type]\n    5 |     }\n      |     ^\n1 error generated.",
    "status_runtime": "N/A",
    "memory": 0,
    "code_answer": [],
    "code_output": [],
    "std_output_list": [
        ""
    ],
    "task_finish_time": 1772310369249,
    "task_name": "judger.runcodetask.RunCode",
    "total_correct": null,
    "total_testcases": null,
    "runtime_percentile": null,
    "status_memory": "N/A",
    "memory_percentile": null,
    "pretty_lang": "C++",
    "submission_id": "runcode_1772310369.0038364_yjJCwK03tT",
    "status_msg": "Compile Error",
    "state": "SUCCESS"
}
```

Run Successfully but wrong answer
```
{
    "status_code": 10,
    "lang": "python3",
    "run_success": true,
    "status_runtime": "46 ms",
    "memory": 19464000,
    "code_answer": [
        "0",
        "2",
        ""
    ],
    "code_output": [],
    "std_output_list": [
        "",
        "",
        ""
    ],
    "elapsed_time": 62,
    "task_finish_time": 1772310248119,
    "task_name": "judger.runcodetask.RunCode",
    "expected_status_code": 10,
    "expected_lang": "java",
    "expected_run_success": true,
    "expected_status_runtime": "73",
    "expected_memory": 42332000,
    "expected_display_runtime": "0",
    "expected_code_answer": [
        "1",
        "3",
        ""
    ],
    "expected_code_output": [],
    "expected_std_output_list": [
        "",
        "",
        ""
    ],
    "expected_elapsed_time": 96,
    "expected_task_finish_time": 1772309750072,
    "expected_task_name": "judger.interprettask.Interpret",
    "correct_answer": false,
    "compare_result": "00",
    "total_correct": 0,
    "total_testcases": 2,
    "runtime_percentile": null,
    "status_memory": "19.5 MB",
    "memory_percentile": null,
    "pretty_lang": "Python3",
    "submission_id": "runcode_1772310247.9741764_jgcGVQ52s9",
    "status_msg": "Accepted",
    "state": "SUCCESS"
}
```

Memory limit exceeded
```
{
    "status_code": 12,
    "lang": "python3",
    "run_success": false,
    "status_runtime": "N/A",
    "memory": 981980000,
    "code_answer": [],
    "code_output": [],
    "std_output_list": [
        ""
    ],
    "elapsed_time": 2506,
    "task_finish_time": 1772311058793,
    "task_name": "judger.runcodetask.RunCode",
    "total_correct": null,
    "total_testcases": null,
    "runtime_percentile": null,
    "status_memory": "N/A",
    "memory_percentile": null,
    "pretty_lang": "Python3",
    "submission_id": "runcode_1772311056.1985083_oygxaSSBaL",
    "status_msg": "Memory Limit Exceeded",
    "state": "SUCCESS"
}
```

Compile error
```
{
    "status_code": 20,
    "lang": "cpp",
    "run_success": false,
    "compile_error": "Line 4: Char 17: error: expected variable name or 'this' in lambda capture list",
    "full_compile_error": "Line 4: Char 17: error: expected variable name or 'this' in lambda capture list\n    4 |         return [10]\n      |                 ^",
    "status_runtime": "N/A",
    "memory": 0,
    "code_answer": [],
    "code_output": [],
    "std_output_list": [
        ""
    ],
    "task_finish_time": 1772311345584,
    "task_name": "judger.runcodetask.RunCode",
    "total_correct": null,
    "total_testcases": null,
    "runtime_percentile": null,
    "status_memory": "N/A",
    "memory_percentile": null,
    "pretty_lang": "C++",
    "submission_id": "runcode_1772311345.4713085_GWXs56Jk8J",
    "status_msg": "Compile Error",
    "state": "SUCCESS"
}
```

Invalid test case
```
{
    "status_code": 15,
    "lang": "python",
    "run_success": false,
    "code_answer": [],
    "code_output": [],
    "runtime_error": "[[\"1\",\"1\",\"1\",\"1\",\"0\"],[\"1\",\"1\",\"0\",\"1\",\"0\"],[\"1\",\"1\",\"0\",\"0\",\"0\"],[\"0\",\"0\",\"0\",\"0\",\"0\"] is not a valid value of type character[][]",
    "full_runtime_error": "[[\"1\",\"1\",\"1\",\"1\",\"0\"],[\"1\",\"1\",\"0\",\"1\",\"0\"],[\"1\",\"1\",\"0\",\"0\",\"0\"],[\"0\",\"0\",\"0\",\"0\",\"0\"] is not a valid value of type character[][]",
    "invalid_testcase": true,
    "case_idx": 2,
    "param_idx": 0,
    "task_finish_time": 1772316729849,
    "task_name": "judger.runcodetask.RunCode",
    "total_correct": null,
    "total_testcases": null,
    "status_runtime": "N/A",
    "runtime_percentile": null,
    "status_memory": "N/A",
    "memory_percentile": null,
    "pretty_lang": "Python",
    "submission_id": "runcode_1772316729.8392327_e7b9ljgkln",
    "status_msg": "Runtime Error",
    "state": "SUCCESS"
}

```

Runtime error
```
{
    "testResultStatus": "Runtime Error",
    "testResult": [],
    "errorMessage": "Line 4: Char 18: runtime error: division by zero (solution.cpp)\nSUMMARY: UndefinedBehaviorSanitizer: undefined-behavior prog_joined.cpp:13:18",
    "lastTestCaseRun": 0,
    "selected": true
}
```
